master/buildbot/util/gceclientservice.py

```diff
@@ -0,0 +1,196 @@
+            req_kwargs['headers']['Authorization'] = "Bearer {0}".format(token)
+
+        return url, req_kwargs
+
+    def zoneEndpoint(self, method=None):
+        base = "/compute/v1/projects/{0}/zones/{1}".format(
+            self.project, self.zone)
+        if method:
+            return "{0}/{1}".format(base, method)
+        else:
+            return base
+
+    def instanceEndpoint(self, method=None):
+        rel = "instances/{0}".format(self.instance)
+        if method:
+            rel = "{0}/{1}".format(rel, method)
+        return self.zoneEndpoint(rel)
+
+    def processRequest(self, deferred):
+        return self.validateRes(deferred)
+
+    @defer.inlineCallbacks
+    def processAsyncRequest(self, deferred):
+        op = yield self.validateRes(deferred)
+        yield self.waitOperationEnd(op)
+
+    def getInstanceState(self, fields=None):
+        if fields is None:
+            return self.get(self.instanceEndpoint())
+        else:
+            return self.get(self.instanceEndpoint(),
+                params={'fields': fields})
+
+    def instanceStart(self):
+        return self.post(self.instanceEndpoint("start"))
+
+    def instanceStop(self):
+        return self.post(self.instanceEndpoint("stop"))
+
+    def createDisk(self, image=None, name=None, type=None):
+        return self.post(self.zoneEndpoint("disks"),
+            json={
+                "sourceImage": image,
+                "name": name,
+                "type": "projects/{0}/zones/{1}/diskTypes/{2}".format(
+                    self.project, self.zone, type)
+            })
+
+    def attachDisk(self, name=None, boot=True, index=0):
+        return self.post(
+            self.instanceEndpoint("attachDisk"),
+            json={
+                "boot": boot,
+                "source": self.zoneEndpoint("disks/{0}".format(name)),
+                "deviceName": name,
+                "index": index
+            }
+        )
+
+    def detachDisk(self, name):
+        return self.post(
+            self.instanceEndpoint("detachDisk"),
+            params={"deviceName": name}
+        )
+
+    def deleteDisk(self, disk_name):
+        return self.delete(
+            "{0}/disks/{1}".format(self.zoneEndpoint(), disk_name))
+
+    def setMetadata(self, fingerprint=None, items=[]):
+        return self.post(self.instanceEndpoint("setMetadata"), json={
+            "fingerprint": fingerprint,
+            "items": items,
+            "kind": "compute#metadata"
+        })
+
+    @defer.inlineCallbacks
+    def waitInstanceState(self, desiredState):
+        state = yield self.processRequest(self.getInstanceState(fields='status'))
+        while state['status'] != desiredState:
+            time.sleep(0.1)
+            state = yield self.processRequest(self.getInstanceState(fields='status'))
+
+    @defer.inlineCallbacks
+    def waitOperationEnd(self, op):
+        while op['status'] != 'DONE':
+            time.sleep(0.1)
```

Comment:
this is a fast polling. won't you quickly reach API quotas with that?

Reply:
Good point. I've changed the sleep time to 0.5 (the API reply takes 0.3 from within Google's own VMs), so that's roughly a ~0.8s query period.

In addition, the code now retries queries on rate-limit errors, with an exponential backoff